src/gc/unix/cgroup.cpp

```diff
@@ -112,16 +121,149 @@ class CGroup
     }
 
     static bool GetCpuLimit(uint32_t *val)
+    {
+        if (GetCGroup1CpuLimit(val))
+        {
+            return true;
+        }
+
+        if (GetCGroup2CpuLimit(val))
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+private:
+    static bool IsCGroup1MemorySubsystem(const char *strTok){
+        return strcmp("memory", strTok) == 0;
+    }
+
+    static bool IsCGroup1CpuSubsystem(const char *strTok){
+        return strcmp("cpu", strTok) == 0;
+    }
+
+    static bool GetCGroup2CpuLimit(uint32_t *val)
+    {
+        char *line = nullptr;
+        char *filename = nullptr;
+        FILE *file = nullptr;
+        char *max_quota_string = nullptr;
+        char *period_string = nullptr;
+
+        size_t lineLen = 0;
+        int sscanRet = 0;
+        long long max_quota = 0;
+        long long period = 0;
+        double cpu_count = 0;
+
+        bool result = false;
+
+        if (s_cpu_cgroup_path == nullptr)
+        {
+            return false;
+        }
+
+        size_t filename_len = strlen(s_cpu_cgroup_path) + strlen(CGROUP2_CPU_MAX_FILENAME);
+        filename = (char*)malloc(filename_len + 1);
+        if (filename == nullptr)
+        {
+            goto done;
+        }
+
+        strcpy(filename, s_cpu_cgroup_path);
+        strcat(filename, CGROUP2_CPU_MAX_FILENAME);
+
+        file = fopen(filename, "r");
+        if (file == nullptr)
+        {
+            goto done;
+        }
+
+        if (getline(&line, &lineLen, file) == -1)
+        {
+            goto done;
+        }
```

Comment:
How long these lines can be?  Might be worth considering using `fgets()` and allocating `line` in the stack instead to save a `malloc()`.

Reply:
I am not sure. The format of this line is `$MAX $PERIOD`. The values are in microseconds, and the first value allows the special case of a `max` literal. [For more information, grep for `cpu.max` here.](https://www.kernel.org/doc/Documentation/cgroup-v2.txt) But there doesn't seem to be any explicit hard limits on the values here.

Reply:
So it's probably safe to say that you need the space for two longs in base 10 plus a space?

Unfortunately you can't use `std::numeric_limits` here, but you can employ a trick to calculate an approximation of the buffer size based on the integer width: `3 * sizeof(integer_type)` will give the size for an integer of type `integer_type`, with space for the terminating NUL, when converted to string in base 10 (it may be larger, but never smaller than needed).

So `1 /* for the space */ + 2 * sizeof(long) */` should be sufficient for a line in this file (and to avoid a `malloc` roundtrip).